src/riak_kv_vnode.erl

```diff
@@ -2322,102 +2346,286 @@ new_md_cache(VId) ->
+update_counter(State=#state{counter=CounterState}) ->
+    #counter_state{cnt=Counter0} = CounterState,
+    Counter = Counter0 +  1,
+    maybe_lease_counter(State#state{counter=CounterState#counter_state{cnt=Counter}}).
+
+
+%% @private we can never use a counter that is greater or equal to the
+%% one fysnced to disk. If the incremented counter is == to the
+%% current Lease, we must block until the new Lease is stored. If the
+%% current counter is 80% through the current lease, and we have not
+%% asked for a new lease, ask for one.  If we're less than 80% through
+%% the lease, do nothing.  If not yet blocking(equal) and we already
+%% asked for a new lease, do nothing.
+maybe_lease_counter(State=#state{counter=#counter_state{cnt=Lease, lease=Lease,
+                                                        leasing=true}}) ->
+    %% Block until we get a new lease, or crash the vnode
+    {ok, NewState} = blocking_lease_counter(State),
+    NewState;
+maybe_lease_counter(State=#state{counter=#counter_state{leasing=true}}) ->
+    %% not yet at the blocking stage, waiting on a lease
+    State;
+maybe_lease_counter(State) ->
+    #state{status_mgr_pid=MgrPid, counter=CS=#counter_state{cnt=Cnt, lease=Lease,
+                                                            lease_size=LeaseSize}} = State,
+    %% @TODO (rdb) configurable??
+    %% has more than 80% of the lease been used?
+    CS2 = if (Lease - Cnt) =< 0.2 * LeaseSize  ->
+                  ok = riak_kv_vnode_status_mgr:lease_counter(MgrPid, LeaseSize),
+                  CS#counter_state{leasing=true};
+             ?ELSE ->
+                  CS
+          end,
+    State#state{counter=CS2}.
+
+%% @private by now, we have to be waiting for a lease, or an exit,
+%% from the mgr. Block until we receive a lease. If we get an exit,
+%% retry a number of times. If we get neither in a reasonable time,
+%% return an error.
+-spec blocking_lease_counter(#state{}) ->
+                                    {ok, #state{}} |
+                                    counter_lease_error().
+blocking_lease_counter(State) ->
+    {MaxErrs, MaxTime} = get_counter_wait_values(),
+    blocking_lease_counter(State, {0, MaxErrs, MaxTime}).
+
+-spec blocking_lease_counter(#state{}, {Errors :: non_neg_integer(),
+                                        MaxErrors :: non_neg_integer(),
+                                        TimeRemainingMillis :: non_neg_integer()}
+                            ) ->
+                                    {ok, #state{}} |
+                                    counter_lease_error().
+blocking_lease_counter(_State, {MaxErrs, MaxErrs, _MaxTime}) ->
+    {error, counter_lease_max_errors};
+blocking_lease_counter(State, {ErrCnt, MaxErrors, MaxTime}) ->
+    #state{idx=Index, vnodeid=VId, status_mgr_pid=Pid, counter=CounterState} = State,
+    #counter_state{lease_size=LeaseSize} = CounterState,
+    Start = os:timestamp(),
+    receive
+        {'EXIT', Pid, Reason} ->
```

Comment:
We would need to call `process_flag(trap_exit, true)` in init/1 to get an `EXIT` message.
See "Receiving Exit Signals" [here](http://www.erlang.org/doc/reference_manual/processes.html)

Are we sure we want to actually trap exits in the vnode though? @jonmeredith @jtuple